orttraining/orttraining/core/graph/optimizer_builder.cc

```diff
@@ -16,5 +16,16 @@ void OptimizerBuilderRegistry::RegisterBuilders() {
   GetInstance().Register<SGDOptimizerBuilder>("SGDOptimizer");
 }
 
+Status IsMatchingTypeAndShape(
+    const onnxruntime::Tensor& tensor,
+    const int32_t element_type,
+    const std::vector<int64_t>& expected_shape) {
+  ORT_RETURN_IF_NOT(tensor.GetElementType() == element_type);
+  const std::vector<int64_t>& tensor_shape = tensor.Shape().GetDims();
+  ORT_RETURN_IF_NOT(tensor_shape.size() == expected_shape.size());
```

Comment:
FYI, vector operator==() should handle differently sized arguments in constant time, so the tensor_shape == expected_shape is probably fine by itself.
https://en.cppreference.com/w/cpp/container/vector/operator_cmp